File: Data Structures & Algorithms/count-paths/submission-2.cpp
class Solution {
public:
    vector<vector<int>> memo;
    int uniquePaths(int m, int n) {
        memo.resize(m, vector<int>(n, -1));
        return dfs(m-1, n-1);
    }
    int dfs(int i, int j) {
        if (i == 0 && j == 0) return 1;
        if (i < 0 || j < 0) return 0;
        if (memo[i][j] != -1) return memo[i][j];
        memo[i][j] = dfs(i-1, j) + dfs(i, j-1);
        return memo[i][j];
    }
};

File: Data Structures & Algorithms/course-schedule-iv/submission-2.py
class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        mp = {i : [] for i in range(numCourses)}
        for pre, crs in prerequisites:
            mp[crs].append(pre)
        memo = {}
        def dfs(crs, pre):
            if pre == crs:
                return True
            if not mp[crs]:
                return False
            key = (crs, pre)
            if key in memo:
                return memo[key]
            for ncrs in mp[crs]:
                if dfs(ncrs, pre):
                    memo[key] = True
                    return True
            memo[key] = False
            return memo[key]
        res = []
        for pre, crs in queries:
            res.append(dfs(crs, pre))
        return res